### ISSUE TITLE
Display works, nothing else

### ISSUE THREAD

**Issue by menime8** [visualapproach/WiFi-remote-for-Bestway-Lay-Z-SPA#330]
Hi there, 
i have a Dreamstream [SPA](https://m.bestwaycorp.at/Product/Detail/1060151NOX22), which has exact the same electronics like Lay Z Spa. Both selled from Bestway. 
My Pump looks like the newer 2021 model.

![IMG_20220710_172051](https://user-images.githubusercontent.com/109034011/178432057-d4f8d43e-c99e-4eaa-b3ad-3238da0c4aee.jpg)
![IMG_20220710_132059](https://user-images.githubusercontent.com/109034011/178432166-947eb7b4-fd21-42f4-af65-ca1343c00f91.jpg)
![IMG_20220710_154049](https://user-images.githubusercontent.com/109034011/178432192-ad282023-30cc-4a28-b010-cf1731bd1128.jpg)
![IMG_20220710_172102](https://user-images.githubusercontent.com/109034011/178432220-02042646-404d-4678-b9e1-4f2de7f344bf.jpg)

I started with PCB V2, NodeMCUV3 and the TXS0108E. First try, nothing happend. 
Second try with 10K resistor, it Starts, DSP starts flashing.
Third try: 560 ohm resistors between DSP and ESP. Display works fine, melody and IP shown. 

But i can't get any connection to the CIO. Wires and connections are all checked with my meter.

I hope anyone can give me a hint 🤔

Thanks

**Comment by menime8** [comment]
> Maybe try resistors btwn LLC and cio also? Or switch to the MOSFET version of LLC. Or try without LLC which is probably fine in practice but may blow the esp in theory according to some people. For the price I'd take the chance...

Tried both versions. Resistors everywhere, then only ESP direct to CIO and DSP. 
![Uploading IMG-20220710-WA0011.jpeg…]()

**Comment by SigmaPic** [comment]
> I have 6 wires, +5v, 0V, RX, TX and one for Sound. What does the last wire?

It's a 3 wire SPI line. Isn't it ?
So, data, clk and chip select.

If you still have access to the pump PCB.
May you tell me if you can read the reference of that chip 

![136789766-c43aaba9-92ff-4629-9687-ae5ffebc5394~2.jpg](https://user-images.githubusercontent.com/2431226/178542336-39d75336-99bf-41b5-b84f-20a1f810fa5d.jpg)

**Comment by menime8** [comment]
I thought it was RX TX.... 
I will have a look tomorrow for chipnumbers.

**Comment by visualapproach** [comment]
> > Maybe try resistors btwn LLC and cio also? Or switch to the MOSFET version of LLC. Or try without LLC which is probably fine in practice but may blow the esp in theory according to some people. For the price I'd take the chance...
> 
> 
> 
> Tried both versions. Resistors everywhere, then only ESP direct to CIO and DSP. 
> 
> ![IMG-20220710-WA0011](https://user-images.githubusercontent.com/109034011/178462846-25ad7d2e-b485-445b-b097-88dca4388947.jpeg)
> 
> 
> 
> 

Oh I see now that you connected 5V to the 3V3 input. Remove that blue wire and check that the ESP is not toasted

**Comment by menime8** [comment]
> > > Maybe try resistors btwn LLC and cio also? Or switch to the MOSFET version of LLC. Or try without LLC which is probably fine in practice but may blow the esp in theory according to some people. For the price I'd take the chance...
> > 
> > 
> > Tried both versions. Resistors everywhere, then only ESP direct to CIO and DSP.
> > ![IMG-20220710-WA0011](https://user-images.githubusercontent.com/109034011/178462846-25ad7d2e-b485-445b-b097-88dca4388947.jpeg)
> 
> Oh I see now that you connected 5V to the 3V3 input. Remove that blue wire and check that the ESP is not toasted

It isn´t.

I used the plan from here [https://github.com/visualapproach/WiFi-remote-for-Bestway-Lay-Z-SPA/discussions/312#discussioncomment-3056766](url)

I know it could be toasted, this was my last try on sunday.

**Comment by menime8** [comment]
> > I have 6 wires, +5v, 0V, RX, TX and one for Sound. What does the last wire?
> 
> It's a 3 wire SPI line. Isn't it ? So, data, clk and chip select instead of RX and TX.
> 
> If you still have access to the pump PCB. May you tell me if you can read the reference of that chip
> 
>
![IMG_20220713_124759](https://user-images.githubusercontent.com/109034011/178716650-0f6e7c99-4bd2-4a68-8c79-dc6dd5589194.jpg)
I read Nuvoton MS51FB9AE

https://www.nuvoton.com/products/microcontrollers/8bit-8051-mcus/industrial-8051-series/ms51fb9ae/

**Comment by menime8** [comment]
Today i tried many different options, with and without resistors, with 560 and 680. It wouldn't work. Checked again all connections.
The other mosfet LLC arrives, too.
I see my ESP has resistors between D3 and 3v and D4 and 3v. I ordered a other ESP. More Info i have on the next few days.

**Comment by visualapproach** [comment]
Did you get rid of the blue wire?

**Comment by menime8** [comment]
Yes
![IMG_20220713_152947](https://user-images.githubusercontent.com/109034011/178775918-9b1e93e1-1d3a-4c74-939f-5311aede207e.jpg)

**Comment by visualapproach** [comment]
Good, I would also try some caps on CIO side and use the B port to get a better LPF. And/or overclock the ESP if there is something speed related.

**Comment by visualapproach** [comment]
Ok so both ends has 1K resistors and unknown caps. So we can calculate a range of cap values with the RC formula. Cut off freq should be higher than the SPI freq. The SPI on my pump is 22 kHz IIRC. I reckon 200-1000 pF would do it.
I need to think a bit on this but what you can check with the scope is the clock frequency on the clock line. And measuring the highs and lows voltages.
I'm not an EE so don't expect miracles here but let's try to find this out

**Comment by SigmaPic** [comment]
> I read Nuvoton MS51FB9AE
> 
> https://www.nuvoton.com/products/microcontrollers/8bit-8051-mcus/industrial-8051-series/ms51fb9ae/
> 

Thanks. It's a microcontroller. I will check the pin config capabilities (push pull, open drain, pull up...)


Looking at the signals with a scope may allow you to detect the root cause of the issue. You can look at signal on CIO side but also on esp side and see if something goes wrong.


> I'm pretty sure, it works with that older Nodemcu.

Which one do you talk about ?


@visualapproach Did you implement a sort of test mode ? We can imagine plugging both connectors from the BWC  together. Then you generate some signal in one side and you check that you can read them on other side.
Only the 5V should be added. By doing this, it would allow to check the BWC board.

**Comment by visualapproach** [comment]
That's a great idea @SigmaPic 
I'm on it

**Comment by SigmaPic** [comment]
> That's a great idea @SigmaPic 
> I'm on it

I'm very busy this time. Otherwise I would have propose a PR.

Maybe you can try to generate 101010 at the rate of the SPI and check if you read the correct bits. Then you change the polarity of the inputs and outputs.

**Comment by SigmaPic** [comment]
> I'm pretty sure, it works with that older Nodemcu.

Ok. I see the resistors: 12k pull up on CS and CLK on DSP side.
On CS, it should not be an issue because this pin is set at high level at startup because it's an active low => the pull up is a good think
For the CLK, it may not be an issue because if CS is high, CLK should no be taken into account.

**Comment by menime8** [comment]
> > I'm pretty sure, it works with that older Nodemcu.
> 
> Which one do you talk about ?

I have the version 3.2. i ordered the 3.0 version with the CH340

**Comment by SigmaPic** [comment]
> Yes
> ![IMG_20220713_152947](https://user-images.githubusercontent.com/109034011/178775918-9b1e93e1-1d3a-4c74-939f-5311aede207e.jpg)
> 
> 

On this photo you don't have the resistors on CIO side. Do you confirm that you tried with them ?

**Comment by menime8** [comment]
> > Yes
> > ![IMG_20220713_152947](https://user-images.githubusercontent.com/109034011/178775918-9b1e93e1-1d3a-4c74-939f-5311aede207e.jpg)
> 
> On this photo you don't have the resistors on CIO side. Do you confirm that you tried with them ?

Yes, tried 560 and 680 ohms

**Comment by visualapproach** [comment]
@SigmaPic I wonder if we can/should set the gpio:s to open drain or with pull up resistors. Don't remember what it is now and what can be done, but could be worth a look. When I have time

**Comment by menime8** [comment]
One question, isn't it possible to use the Miso/Mosi/CLK from the ESP?

![IMG_20220714_104431](https://user-images.githubusercontent.com/109034011/178941724-ea228610-9c10-42c9-981f-e4e5fd852858.jpg)

**Comment by visualapproach** [comment]
Espressif states mode 3 is not supported.

**Comment by SigmaPic** [comment]
> I'm pretty sure, it works with that older Nodemcu.

I think it won't change anything because there are some pull up on your mosfet LLC. See my shematic below.
So the pull up on esp side won't change anything.
But, you can try to be sure. Maybe I'm wrong.

> @SigmaPic I wonder if we can/should set the gpio:s to open drain or with pull up resistors. Don't remember what it is now and what can be done, but could be worth a look. When I have time

I take a look at the full schematic and I found something strange that may explain some troubles.
The issue concerns the mosfet LLC when data are sent from DSP/CIO to ESP.
The other way is ok (from ESP to DSP/CIO).

As you can see on  my simulation, the voltage on ESP input stay at 1.5V when DSP/CIO drive the line to 0.
This comes from the serial 1k resistor.
On this simulation, I put one 1k which is the one in the DSP/CIO and one 1k which is the one we put on our PCB.
If I remove one 1k, it's better but not perfect.
To get perfect results, I need to remove both 1k resistors which is not possible because we have at least one.

![shifter](https://user-images.githubusercontent.com/2431226/178977945-92c5a4df-76e8-4d2e-9f8b-82c21017c467.PNG)

Why it happens ?
Because these two 1k resistors act as a voltage divider with 10k pull up on the LLC (5V side).
This means that the input of the LLC is at 0.83V instead of 0V due to this voltage divider.
Because of that, the mosfet is not driven and the output of the LLC is 0.83V + mosfet diode forward volatge which is about 0.8V.
So we get about 1.5/1.6V in output of the LLC which is greater that the input low voltage of the ESP which is 0.25x3.3 = 0.8V.

To confirm that issue, we could developp a dumy test SW that just put all pins in input and print the state in the console.
Then we can manually put 5V on each pin and check if it's correctly detected.

As a workaround, we can eventually modify  the resistors mounted on the LLC.
Put for example 50k instead of 10k.
An removed all pull up resistors on ESP side, the ones we talked previously that are externally mounted on some ESP.
By doing this, we are theoretically in the spec of the chip.
But it requires some soldering and the PCB of the LLC is very cheap => difficult to replace the resistor without breaking the PCB.
![custom_llc](https://user-images.githubusercontent.com/2431226/178985554-3fd384f8-03c8-4f2e-8a79-c3bd35e2cd75.PNG)

To "perfectly" solve that issue the easiest way is to use two pin of the ESP to handle the data line.
For all other lines (CS, CLK, audio), we can continue to the use the mosfet LLC or replace it by simple mosfet as shown on this schematic. But these other lines work in a single direction, so only one mosfet and one pin is needed.
![new_shifter](https://user-images.githubusercontent.com/2431226/178981150-88a9d369-f01c-494a-9cbf-efd6cf15f4ac.PNG)

@visualapproach What do you think about that ?

But now you may wonder why it work for some people ?
It may work for some people because, we reach 1.5/1.6V on the input of the ESP.
According to ESP datasheet, everything under 0.8V is seen as 0 and everything over 2.5V is seen a 1.
It means that everything between 0.8V and  2.5V is undertermined. Depending on some factors, it may work or not.

And why not using the recommended TXS0108E ?
Finally, it may be better to use TXS0108E instead of mosfet LLC. What I don't like is that Texas recommend to shorten the output wires and not use capacitive load. In out case, we have long wires and filter with capacitors. So, I'm not sure that the chip works correctly.

My old school design with discrete mosfets takes two more pins (one for the CIO and one for the DSP) but looks more robust.

**Comment by SigmaPic** [comment]
> ![IMG_20220713_125236](https://user-images.githubusercontent.com/109034011/178812669-49872bcd-6c3d-4fd3-86fe-4e15403f6978.jpg)

About the pinout of the CIO.

Where is the 5V and the GND ?
Pin 5 = GND
Pin 6 = 5V 

??

**Comment by SigmaPic** [comment]
> First I think you've made a fantastic job with the schematics and explanations. Very good! Regarding the double pins per data wire, I abandon it for two reasons; 1 it requires more components and therefore more difficult for most people. Nowadays I think you can get the PCB fabs to populate the PCB but I haven't tried that and it seemed difficult to find the right components. 2 there isn't enough GPIOs on the ESP8266. Not entirely true, but D0 cannot have interrupts and D8 is pulled to GND and boot fails if pulled high. Now, there is RX and TX pins also which I don't want to mess with because it will also make installation harder. However we might pull it off if we use D0 (no irq), D8 (pulled down), D3, D4 (pulled up) and D5 to the DSP since we don't need interrupts there. Leaving D1, D2, D6 and D7 for the CIO. So much for parallel traces :-)
> 

Yes it requires to solder SOT23 chips. That's the price to pay :)
D0 and D8 are perfect, isn't it ?
No need of interrupt on data line, isn't it ?
and D0 can be used as an output for data line with an external pull down resistor if needed.

> There is also the possibility to switch to ESP32 with lots more GPIOs and horsepower to serve web pages quicker. But it consumes more power i guess, which in some cases seem to be a weak spot. And not to mention - I haven't got it to work yet. (Haven't put too much effort in yet but it would take a lot of time recoding, and making new PCB.)

Yes, it will take a lot of effort.

> So, any other options? 

Direct connection.
Maybe we can add a diode from ESP input to VCC so that the ESP is protected.
This resistor will act has a voltage divider with the serial resistors.
We can increase a bit the serial resistor value to reduce the current coming from the CIO.


> Are there LLCs with weaker pullups to buy? For your information, the reason I introduced the TX0108E is because for a while the other one was difficult to source and much more expensive. Now I think there is no real difference.

I don't think you can buy LLC with 50K resistors.

@visualapproach  What is your opinion about @menime8 issue ?
He tried with a direct connection.
It means it's not an issue related to voltgae shifting.

Do you think it's a good idea to print in the console the command received from CIO ?
Every time the CS goes high, you coul print in the console (debug.html) th value of _receivedByte.
This would allow to see if he get some command or not.

**Comment by visualapproach** [comment]
> @visualapproach What is your opinion about @menime8 issue ?
He tried with a direct connection.
It means it's not an issue related to voltgae shifting.

I don't know, there may be something more to it in that case. Don't know if he blew the ESP by shorting 5V and 3V3 or if he has a new one now.

> Do you think it's a good idea to print in the console the command received from CIO ?
Every time the CS goes high, you coul print in the console (debug.html) th value of _receivedByte.
This would allow to see if he get some command or not.

As a one off in this case maybe. Otherwise its a fulltime job to try to decode what's wrong, and it is usually hardware fault caused by bad soldering, bad wires or overheated components during soldering. I think first off is to reduce the chance of this by doing a board self test as you suggested. Then if it does not work for several people maybe go that route. But didn't he have an oscilloscope? Or was it someone else? Easiest would be to check out the signals at the ESP with the scope I think.
You should read something like this:
CS.  `----__________________________________------`
CLK:`-----_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-------`
DAT:`-----___-_---_--____--____--_-____---------`

**Comment by SigmaPic** [comment]
> > @visualapproach What is your opinion about @menime8 issue ?
> > He tried with a direct connection.
> > It means it's not an issue related to voltgae shifting.
> 
> I don't know, there may be something more to it in that case. Don't know if he blew the ESP by shorting 5V and 3V3 or if he has a new one now.
> 
> > Do you think it's a good idea to print in the console the command received from CIO ?
> > Every time the CS goes high, you coul print in the console (debug.html) th value of _receivedByte.
> > This would allow to see if he get some command or not.
> 
> As a one off in this case maybe. Otherwise its a fulltime job to try to decode what's wrong, and it is usually hardware fault caused by bad soldering, bad wires or overheated components during soldering. I think first off is to reduce the chance of this by doing a board self test as you suggested. Then if it does not work for several people maybe go that route. But didn't he have an oscilloscope? Or was it someone else? Easiest would be to check out the signals at the ESP with the scope I think. You should read something like this: CS. `----__________________________________------` CLK:`-----_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-------` DAT:`-----___-_---_--____--____--_-____---------`

Your right, the shortcut may jae broken something.

@visualapproach Do you know if the GND is connected to the ground of the main power ?
In other word, is it possible to connect the GND of the CIO to the mass of the scope without any electrical issue ?
Sometimes, the GND can be referenced to the main power (230V) and connecting the mass of the scope to the GND can provoke a short circuit. No issue here ?

**Comment by visualapproach** [comment]
@SigmaPic You are right, I don't know where the GND goes to so maybe a bad idea. Please do not destroy your scopes.

**Comment by visualapproach** [comment]
And also @air101 found out that his problem was due to a particular type of NodeMCU. #325

**Comment by menime8** [comment]
Yes, it was me with the osci. Tomorrow the other ESP arrives. I another issue the ESP was faulty. I think there are a few fakes on the market. I'm very approved with the soldering iron, also with hot air for SMD. I repair since 2015 car amplifiers. Since then i don't have one failed solderpoint. Also i double checked every connection with the meter. 

Before the bridge 3 to 5 Volt i also tried ever possible constellation. 

I use a USB scope with a laptop! Hantek 6022BE.
So i isolated. 
If the 0V is grounded, the RCD should trigger, i touched it few times while measuring.

Lets see what i can find at the next weekend.

**Comment by visualapproach** [comment]
@menime8 Sorry for talking about you in third person. The statement about bad solders was a general experience, not directed at you. I just described my situation where a lot of time searching for code errors often ends with success after mentioned problems had been fixed. I think you are great trying to help people in the forums!

**Comment by menime8** [comment]
@visualapproach 

No problem. Your code works, i think this is approved more than 100 times. My connections are fine, too. So i see the issue at the nodeMCU. My Version is 3.2 with the ESP8266 12F, they say its the better WiFi Version. Now i ordered the older 12E.

**Comment by visualapproach** [comment]
@SigmaPic I just uploaded a new version to test the setup by connecting the cables together. It runs 1000 edges on each data line, both directions and gives back a report. You need to visit the page "/hwtestinfo.html" to run the test.

**Comment by SigmaPic** [comment]
> @SigmaPic I just uploaded a new version to test the setup by connecting the cables together. It runs 1000 edges on each data line, both directions and gives back a report. You need to visit the page "/hwtestinfo.html" to run the test.

Nice !!!
Thanks

@menime8 Do you want to test this software. It may allow you to confirm that's there's is an hardware problem on BWC side

**Comment by SigmaPic** [comment]
> So, any other options? Are there LLCs with weaker pullups to buy? For your information, the reason I introduced the TX0108E is because for a while the other one was difficult to source and much more expensive. Now I think there is no real difference.

After thinking a bit more about that topic, I think  I have found "the" easiest solution that fully matches all specifications. I have to check the datasheet again and simulate it but on the paper it works.
And the the most amazing is that it can be implemented ok the PCB V2.
The solution is so easy that I wonder if I missed something.
Details for tomorrow. ;)

**Comment by visualapproach** [comment]
Curious:)

**Comment by SigmaPic** [comment]
The idea is to use a simple voltage divider to convert the 5V coming from CIO/DSP into 3.3V.
In the other way, no need of conversion since 3.3V is seen as a logical 1 by the CIO/DSP.

Signals look perfect and it could almost be implemented on PCB v2.

![llc_with_res](https://user-images.githubusercontent.com/2431226/179343927-44336016-c48f-489a-953e-36150f6baff3.PNG)

The only issue is the pull up resistor on D3/D4 that make the mess.

These pins shall be replaced by D0/D8 or something else as you suggested it.
That's really annoying, there are not so much pins and 3 of them shall be pulled up or down.


> Not entirely true, but D0 cannot have interrupts and D8 is pulled to GND and boot fails if pulled high.

We could also use D11 and D12, isn't it ?

@visualapproach What do you think ?

It's easier and cheaper. No need of LLC.

Do you have an idea why the DSP is flashing at boot and why adding serial resistor helps ?

**Comment by visualapproach** [comment]
> The only issue is the pull up resistor on D3/D4 that make the mess.

I think one of them is for the on board LED. How much do they affect the signal?

> We could also use D11 and D12, isn't it ?

What is that? I'm only aware of D0-D8

> Do you have an idea why the DSP is flashing at boot and why adding serial resistor helps ?

Not really. Obviously random but errors would case things to behave weird. There is 'enable-bits' that would flash the display if toggled, but the strange thing is that it works during start up sequence.

I think it's a nice idea. Would need some mods on the pcb to make it pretty and easy. I wonder if it would work on all other models as well. Possibly the audio will be more silent but it's no good on the 2021 models anyway from what I have heard. What was your setup again @SigmaPic ? Is it working for you?

**Comment by menime8** [comment]
> > The only issue is the pull up resistor on D3/D4 that make the mess.
> 
> I think one of them is for the on board LED. How much do they affect the signal?

I deleted the 2 resistors, another idiot idea :-) LED is working. 

Where i can find the new software? I think i´m blind...

My new ESP arrived 2 hours ago.

**Comment by menime8** [comment]
![err](https://user-images.githubusercontent.com/109034011/179355760-e1f5d869-97d0-4700-84be-4d545780e08f.png)

There missing files i need?

**Comment by visualapproach** [comment]
@menime8 Newest version is found in the development branch:
![image](https://user-images.githubusercontent.com/22987250/179355919-dcc36d54-6637-48ad-b7bf-f3c65c5094a4.png)

Sometimes one need to clean build files:
![image](https://user-images.githubusercontent.com/22987250/179355962-50ec7471-be35-4812-965e-052538d68bbb.png)

**Comment by menime8** [comment]
> /hwtestinfo.html

DSP input pin 2 0 errors of 1000
DSP input pin 3 0 errors of 1000
DSP input pin 4 0 errors of 1000
CIO input pin 2 0 errors of 1000
CIO input pin 3 0 errors of 1000
CIO input pin 4 0 errors of 1000

If i start the test with disconnected plugs i get:

DSP input pin 2 500 errors of 1000
DSP input pin 3 500 errors of 1000
DSP input pin 4 500 errors of 1000
CIO input pin 2 500 errors of 1000
CIO input pin 3 500 errors of 1000
CIO input pin 4 500 errors of 1000

**Comment by visualapproach** [comment]
Nice. 500 is = no contact.

**Comment by menime8** [comment]
> Nice. 500 is = no contact.

Also test with single broken contacts, works fine. Thank you. 

Lets see what my CIO says

**Comment by SigmaPic** [comment]
> I think one of them is for the on board LED. How much do they affect the signal?

You're right, the LED is on D0.
Maybe D11 and D12 is better.

> > We could also use D11 and D12, isn't it ?
> 
> What is that? I'm only aware of D0-D8

D11 and D12 is located on the other side of the esp. Look on the schematic [https://github.com/nodemcu/nodemcu-devkit-v1.0/blob/master/NODEMCU_DEVKIT_V1.0.PDF]()


> > Do you have an idea why the DSP is flashing at boot and why adding serial resistor helps ?
> 
> Not really. Obviously random but errors would case things to behave weird. There is 'enable-bits' that would flash the display if toggled, but the strange thing is that it works during start up sequence.

Sorry, I didn't catch you. What are the 'enable-bits' you talked about?
Can you explain ?


> I think it's a nice idea. Would need some mods on the pcb to make it pretty and easy. I wonder if it would work on all other models as well. 

Yep it depends on the microcontroller of the other models. But it should work I think. I will check again that point.

> Possibly the audio will be more silent but it's no good on the 2021 models anyway from what I have heard. 

What is no good on 2021 model ?
I am interested because I have a 2021.


> What was your setup again @SigmaPic ? Is it working for you?

I have ordered a PCB V2 customized by me with SMD resistors and a jumper to root the audio directly from CIO to DSP. I can share it once it will be tested. I have a model 2021.

**Comment by SigmaPic** [comment]
> > /hwtestinfo.html
> 
> DSP input pin 2 0 errors of 1000
> DSP input pin 3 0 errors of 1000
> DSP input pin 4 0 errors of 1000
> CIO input pin 2 0 errors of 1000
> CIO input pin 3 0 errors of 1000
> CIO input pin 4 0 errors of 1000
> 
> If i start the test with disconnected plugs i get:
> 
> DSP input pin 2 500 errors of 1000
> DSP input pin 3 500 errors of 1000
> DSP input pin 4 500 errors of 1000
> CIO input pin 2 500 errors of 1000
> CIO input pin 3 500 errors of 1000
> CIO input pin 4 500 errors of 1000
> 
> AND: with PCBV2 you don´t need 5V, because you´ll get it from your USB-Port. It is easier for the most people.

Dis you try with the old esp ?

**Comment by SigmaPic** [comment]
> I deleted the 2 resistors, another idiot idea :-) LED is working. 

If you use a MOSFET LLC, you can remove it because there are some pull ups on the LLC.

**Comment by menime8** [comment]
> Dis you try with the old esp ?

With old and new, same the same results.
So now the new one also won't work.

Here i have a few shots with the osci. Attention, different time dividers.
![IMG_20220716_192332](https://user-images.githubusercontent.com/109034011/179365745-59a5d8ad-f2cc-4a9f-8a30-1ddf24297279.jpg)
![IMG_20220716_192302](https://user-images.githubusercontent.com/109034011/179365765-497beaac-5629-4a35-be28-c95ad84d2099.jpg)
![Uploading IMG_20220716_192237.jpg…]()

**Comment by RatAttack83** [comment]
Interesting news in this topic... So I have disconnectet my spa and tried to get it to work again... with Powerbank and with laptop... it wont work... :( 
I ordered some othe NodeMCU and tried it... same result...
so i can measure with a mobile scope, but my contact is on holiday for the next three weeks and my 230V scope is not the best idea.
But i have read a bit in this topic and think you are on the right way. :) i hope you will get the Model 2021/22 to work :)

If someone give me shematics to test, i can etch it at work and check it at home :) so perhaps i can contribute a little.

**Comment by SigmaPic** [comment]
> > Dis you try with the old esp ?
> 
> With old and new, same the same results.
> So now the new one also won't work.
> 
> Here i have a few shots with the osci. Attention, different time dividers.
> ![IMG_20220716_192332](https://user-images.githubusercontent.com/109034011/179365745-59a5d8ad-f2cc-4a9f-8a30-1ddf24297279.jpg)
> ![IMG_20220716_192302](https://user-images.githubusercontent.com/109034011/179365765-497beaac-5629-4a35-be28-c95ad84d2099.jpg)
> [
> ![IMG_20220716_192237](https://user-images.githubusercontent.com/109034011/179365794-b7082c39-7905-4c10-8e86-1729592e1a1b.jpg)
> ](url)



I think it's chip select, clock and last one data. Isn't it ?
Clearly the last signal is not correct. Edges are too smooth. Too much filtered.
What was your setup? Which LLC ? Resistors ?

**Comment by menime8** [comment]
> I think it's chip select, clock and last one data. Isn't it ? Clearly the last signal is not correct. Edges are too smooth. Too much filtered. What was your setup? Which LLC ? Resistors ?

Scoped with original DSP on it, nothing else.

**Comment by SigmaPic** [comment]
>  > I think it's chip select, clock and last one data. Isn't it ? Clearly the last signal is not correct. Edges are too smooth. Too much filtered. What was your setup? Which LLC ? Resistors ?
> 
> Scoped with original DSP on it, nothing else.
> 

😳 That's really strange.
It's the first time you look at the signal ?

**Comment by menime8** [comment]
> 
Yes. I will now disassemble the display again, to see what R or RL there is.

**Comment by SigmaPic** [comment]
I decoded the signal,  it's a command 0x42 0x1B 0x1B

Maybe other people have already looked at the signal and know if the waveform is normal.

It's strange because some edges are Ok and some other edges are smooth.

**Comment by menime8** [comment]
![IMG_20220717_120402](https://user-images.githubusercontent.com/109034011/179393357-3a0d822d-26ce-4d41-a2df-767aed344ed3.jpg)
![Screenshot_2022-07-17-12-03-40-196_com android chrome](https://user-images.githubusercontent.com/109034011/179393365-ae17f131-d84e-4425-b1ce-43014619468d.jpg)

Shorten the wires, thanks for your hint with that Strange signal. I think the wire acts like a capacitance.

Now i can read the actual temp and thats it. No switch is working.

**Comment by menime8** [comment]
> ![IMG_20220717_120402](https://user-images.githubusercontent.com/109034011/179393357-3a0d822d-26ce-4d41-a2df-767aed344ed3.jpg) ![Screenshot_2022-07-17-12-03-40-196_com android chrome](https://user-images.githubusercontent.com/109034011/179393365-ae17f131-d84e-4425-b1ce-43014619468d.jpg)
> 
> Shorten the wires, thanks for your hint with that Strange signal. I think the wire acts like a capacitance.
> 
> Now i can read the actual temp and thats it. No switch is working.

I use now the pre2021. I get data from CIO, but can't send any back.

**Comment by menime8** [comment]
> > ![IMG_20220717_120402](https://user-images.githubusercontent.com/109034011/179393357-3a0d822d-26ce-4d41-a2df-767aed344ed3.jpg) ![S
> > Shorten the wires, thanks for your hint with that Strange signal. I think the wire acts like a capacitance.
> > Now i can read the actual temp and thats it. No switch is working.
> 
> I use now the pre2021. I get data from CIO, but can't send any back.

**Comment by SigmaPic** [comment]
@menime8  Can you monitor the signal on the input of the esp.

If you can do it with the LLC as well. I think we will understand why it doesn't work.

**Comment by menime8** [comment]
> @menime8 Can you monitor the signal on the input of the esp.
> 
> If you can do it with the LLC as well. I think we will understand why it doesn't work.

Lets see if i would do this next few days. Everytime disassemble everything and bring the pump inside,  sometimes i would use the pool 😉
And it needs filtered every day.

**Comment by SigmaPic** [comment]
What change with pre2021 ?

It's the pinout ?

Can you tell us the correspondence between the scope screenshot and the pin number on the connector ?

**Comment by visualapproach** [comment]
Have you tried MIAMI2021? It's for the newer pumps. It uses different codes for the buttons etc. Or maybe the special version 54149E that uses two data lines and a clock.

**Comment by visualapproach** [comment]
> I decoded the signal, it's a command 0x42 0x1B 0x1B
> 
> Maybe other people have already looked at the signal and know if the waveform is normal.
> 
> It's strange because some edges are Ok and some other edges are smooth.

that would be right. the 0x42 is the CIO asking the DSP for which button is pressed. 0x1b1b means no button.

**Comment by menime8** [comment]
> Have you tried MIAMI2021? It's for the newer pumps. It uses different codes for the buttons etc. Or maybe the special version 54149E that uses two data lines and a clock.

I tried all the time Miami, nothing.
Today i tried Pre2021 and i see the temp in the display. But only after i shorten the cables again. Needs resistors between DSP and ESP.
When i solder resistors between CIO and ESP the connection won't work again.

I try the other Software these days.

Here another pictures from DSP PCB

![IMG_20220717_112607](https://user-images.githubusercontent.com/109034011/179400347-1e1482fc-868e-4608-bfcf-57bed379c52e.jpg)
![IMG_20220717_112743](https://user-images.githubusercontent.com/109034011/179400364-5d461418-9f3e-4a69-9b0d-554fef805475.jpg)

**Comment by menime8** [comment]
> What change with pre2021 ?
I'm interested, too.

What is the difference? Other Hex-Codes or other SPI config?

**Comment by visualapproach** [comment]
Same chip as in my display:
#36

And yes, same SPI but different bitpatterns. It appears from your scope pics that it's clocking faster than mine. 40KHz vs 22?

**Comment by menime8** [comment]
> 

I'll have a closer look tomorrow, about the frequencies.

**Comment by RatAttack83** [comment]
> ![IMG_20220717_120402](https://user-images.githubusercontent.com/109034011/179393357-3a0d822d-26ce-4d41-a2df-767aed344ed3.jpg) ![Screenshot_2022-07-17-12-03-40-196_com android chrome](https://user-images.githubusercontent.com/109034011/179393365-ae17f131-d84e-4425-b1ce-43014619468d.jpg)
> 
> Shorten the wires, thanks for your hint with that Strange signal. I think the wire acts like a capacitance.
> 
> Now i can read the actual temp and thats it. No switch is working.

Same here... Temp is at MQTT, too but nothing else... Pushbutton makes a beep but takes no affect. Also via MQTT 😢 
But the Webserver only starts safe if i connect a powerbank, without no connection.

**Comment by menime8** [comment]
> Same here... Temp is at MQTT, too but nothing else... Pushbutton makes a beep but takes no affect. Also via MQTT 😢 But the Webserver only starts safe if i connect a powerbank, without no connection.

Can you post a picture from your PCB? Tried the 54149E Version? Which pump model?

**Comment by visualapproach** [comment]
You could rightclick in the browser and find inspect/console. Then check "button codes" when you press different buttons on the display. Compare with the defines in libs/bwc/const.h (from the top of my head. Not at computer)
Then you will be able to determine which model to use.

**Comment by SigmaPic** [comment]
> Here another pictures from DSP PCB
> 
> ![IMG_20220717_112607](https://user-images.githubusercontent.com/109034011/179400347-1e1482fc-868e-4608-bfcf-57bed379c52e.jpg)
> ![IMG_20220717_112743](https://user-images.githubusercontent.com/109034011/179400364-5d461418-9f3e-4a69-9b0d-554fef805475.jpg)
> 

I looked at the datasheet of the TM1628 and I found something very interesting.
They explain that DIO pin is open drain pin which is plausible. And they recommend to put a pull up on that pin. On the recommended schematic, they put pull up on all signals.

Where are located these pull ups ?
On CIO or on DSP board.

On the CIO, we can see some via that go somewhere but I don't know where.

@visualapproach are you aware about that ?

If there is no pull up on DSP side, it means that these pull up shall be on BWC. 
With an LLC, no problem because they are present on the LLC but with a direct connection it can be unstable.

**Comment by visualapproach** [comment]
@SigmaPic I'm not aware of where the pull up resistors are located. It would make more sense to put them on the CIO side as it's providing the 5V in the first place. But not necessary I guess. It is a while since I made this but I think I measured 5V on the data pins from the CIO. Now I have a 54189 (PRE2021) so things may have changed. I think it is strange that there is so many problems/solutions recently. New NodeMCU solved it for one, loosing the LLC for another, resistors for many other, using "the other" LLC for some. I like your way of attacking the problem though.

**Comment by SigmaPic** [comment]
It's because some people have trouble that I search why. I guess we miss something that will make it work for everybody :)

In #116 we can clerly see the pull up resistors. they are on the DSP board:

![image](https://user-images.githubusercontent.com/2431226/179455070-bbf90dfb-7d20-42e9-8c26-6585aaf24345.png)

**Comment by menime8** [comment]
> You could rightclick in the browser and find inspect/console. Then check "button codes" when you press different buttons on the display. Compare with the defines in libs/bwc/const.h (from the top of my head. Not at computer) Then you will be able to determine which model to use.

I can´t get the MODEL54149E to work. It seems like a watchdog error.

the ESP reboots every 10 secs.

Serialmonitor says:
wdt reset
load 0x4010f000, len 3460, room 16
tail 4
chksum 0xcc
load 0x3fff20b8, len 40, room 4
tail 4
chksum 0xc9
csum 0xc9
v000772c0
~ld
Failed to load settings.txt

**Comment by menime8** [comment]
They put the old eggshaped pcbs in the new square housings and sell them less cheaper under a other name. I payed 199€/~200$ for it 😅

**Comment by SigmaPic** [comment]
Your last scope screenshot have been taken without the esp ?

**Comment by menime8** [comment]
> Your last scope screenshot have been taken without the esp ?

No

**Comment by SigmaPic** [comment]
> > Your last scope screenshot have been taken without the esp ?
> 
> No

What setup ? What resistor ?

**Comment by menime8** [comment]
> > > Your last scope screenshot have been taken without the esp ?
> > 
> > 
> > No
> 
> What setup ? What resistor ?

Original with DSP. So now someone can say which is data and which clock. Last time i measured, but did not noticed which picture from which wire was.

**Comment by visualapproach** [comment]
if you have a PRE2021 and press UNIT(C/F) button on the display (may have to keep in pressed a few seconds) you will see code '1012'. If you see '0809' you have a MIAMI2021. If you see '0a09' you have a maldives2021 with hydrojets.

**Comment by visualapproach** [comment]
pin 3 is data, 
pin4 is clock, 
pin 5 is chip enable/CS

All signals look good except data which looks like a capacitance issue as @SigmaPic already said. Or what do you make of it @SigmaPic ?

**Comment by visualapproach** [comment]
Did you try @160MHz @menime8 ?

**Comment by menime8** [comment]
> Did you try @160MHz @menime8 ?

No. Tomorrow the next and i think the last try 😅
I'll try 160mhz, i measure again my 22 or 40 kHz and i also do so some shots with ESP in data line

**Comment by menime8** [comment]
> > > if you have a PRE2021 and press UNIT(C/F) button on the display (may have to keep in pressed a few seconds) you will see code '1012'. If you see '0809' you have a MIAMI2021. If you see '0a09' you have a maldives2021 with hydrojets.
> > 
> > 
> > I See 0200, 1312, 1112, 1212, 1012, 0300 and 0100, possible i tried the unit button NOT.
> 
> You need to check which button is what because those values are used in all models, just in a different order. The UNIT button would be the easiest check because it differs between the models.

I'll check it.

**Comment by SigmaPic** [comment]
Ok nice !

So, it's the same pinout that the one implemented in ESP code:
Pin 1 is +5v
Pin 2 is GND
Pin 3 Data
Pin 4 CLK
Pin 5 CS

What is interesting for me is that it confirm was I discovered about the TM1628 that is an open drain on data pin.
You screenshots confirm that point and let me think that CIO is in push pull because:

1. When the CIO send data, the low level is not 0V but 0.8V and edges are sharp
2. When the DSP send data the low level is 0V but edges are smooth

This is a very strange hardware design.

![image](https://user-images.githubusercontent.com/2431226/179569925-3488075d-e5c3-44bc-a782-77afcfc817b5.png)

I will think about that and suggest a new way of voltage conversion.

**Comment by visualapproach** [comment]
Awesome @SigmaPic ! I think it's worth a try. Your calculations has more chance of success than my guesses 😃 
I can make a new PCB define in the code and then me or you make a drawing how to connect the cables. As I see it you need to make a braid almost, but that's cool. If it works we can do a new PCB later.

**Comment by visualapproach** [comment]
OK @menime8 @RatAttack83 @SigmaPic , so I uploaded new version (DEV BRANCH). Edit model.h and use PCB_V2B. Reconnect wires as described above and in the comments in model.h. Use red LLC.

**Comment by SigmaPic** [comment]
> Awesome @SigmaPic ! I think it's worth a try. Your calculations has more chance of success than my guesses smiley I can make a new PCB define in the code and then me or you make a drawing how to connect the cables. As I see it you need to make a braid almost, but that's cool. If it works we can do a new PCB later.

> OK @menime8 @RatAttack83 @SigmaPic , so I uploaded new version (DEV BRANCH). Edit model.h and use PCB_V2B. Reconnect wires as described above and in the comments in model.h. Use red LLC.

Nice !!

This is the temporary schematic.
Once we know it work, I can update the easyEDA schematic and even the PCB if you want.

Do you confirm my schematic ?

![image](https://user-images.githubusercontent.com/2431226/180045211-a946b9e3-13a9-4e7a-a91a-20de6f7b675d.png)

**Comment by menime8** [comment]
@visualapproach @SigmaPic 
Thanks for the great work, i'll direct try this tomorrow morning.

**Comment by visualapproach** [comment]
@SigmaPic 
> This is the temporary schematic.
Once we know it work, I can update the easyEDA schematic and even the PCB if you want.

That would be nice! 

> Do you confirm my schematic ?
The pinout seems right, but B-ports are not connected to the rest of the board unless you put resistors or jumpers in. So maybe make the drawings to the A-ports and if someone needs to add resistors they should move those pins to B-port.
(The caps will still work if connected to A-ports since they just go to GND.)

**Comment by SigmaPic** [comment]
You're right, schematic updated.

I will wait that someone confirm that it works before updating the easyEDA schematic

**Comment by menime8** [comment]
> You're right, schematic updated.
> 
> I will wait that someone confirm that it works before updating the easyEDA schematic

The DEV code doesn´t work. The #elif needs a "define()" like this:

#ifdef PCB_V2
//DATA, CLK, CS
const int ciopins[] = {D1, D2, D3};
const int dsppins[] = {D4, D5, D6, D7};
#elif defined(PCB_V2B)
const int ciopins[] = {D1, D2, D5};
const int dsppins[] = {D6, D4, D3, D7};
#else
const int ciopins[] = {D7, D2, D1};
const int dsppins[] = {D5, D4, D3, D6};
#endif

**Comment by Bischof-Mak** [comment]
i have the same error like you


      |  ^~~~~~~
In file included from lib\BWC/BWC_common.h:22,
                 from src\main.h:5,
                 from src\main.cpp:1:
lib\BWC/BWC_6w_type1.h:11:14: error: #elif with no expression
   11 | #elif PCB_V2B

      |              ^

**Comment by menime8** [comment]
> i have the same error like you
> 
> ```
>   |  ^~~~~~~
> ```
> 
> In file included from lib\BWC/BWC_common.h:22, from src\main.h:5, from src\main.cpp:1: lib\BWC/BWC_6w_type1.h:11:14: error: #elif with no expression 11 | #elif PCB_V2B
> 
> ```
>   |              ^
> ```

Write this in line 11: #elif defined(PCB_V2B)
instead #elif PCB_V2B.


My connections Work. Everything is working, im very Happy ATM. Thanks to @visualapproach and @SigmaPic .

So you have a new Model for your list.

I have now this 
![16583961231655484991716261430678](https://user-images.githubusercontent.com/109034011/180181919-e95b13a8-357f-4f1a-844a-e8186b61fc6d.jpg)

 Now i would solder it proper.

**Comment by menime8** [comment]
> > i have the same error like you
> > ```
> >   |  ^~~~~~~
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In file included from lib\BWC/BWC_common.h:22, from src\main.h:5, from src\main.cpp:1: lib\BWC/BWC_6w_type1.h:11:14: error: #elif with no expression 11 | #elif PCB_V2B
> > ```
> >   |              ^
> > ```
> 
> Write this in line 11: #elif defined(PCB_V2B) instead #elif PCB_V2B.
> 
> My connections Work. Everything is working, im very Happy ATM. Thanks to @visualapproach and @SigmaPic .
> 
> So you have a new Model for your list.
> 
> Now i would solder it proper.

I was a bit to early. The Webinterface toggles and switch different buttons on....


https://user-images.githubusercontent.com/109034011/180183161-bdc95f39-9231-4dd2-8a2a-780b65a88c0f.mp4

**Comment by Bischof-Mak** [comment]
You professionals have made it !!!!!


now I managed to load the DEV_Branch.

Module connected, power plugged in, everything boots up, everything can be operated (display and web interface).

Well done.

**Comment by SigmaPic** [comment]
> You professionals have made it !!!!!
> 
> 
> now I managed to load the DEV_Branch.
> 
> Module connected, power plugged in, everything boots up, everything can be operated (display and web interface).
> 
> Well done.

I'm very happy that it works correctly.
Same setup that the photo you sent yesterday?

I can update the schematic and the PCB.

**Comment by SigmaPic** [comment]
@menime8 Maybe it's an issue of button code. I don't remember. Did you already check the button code in the console off the browser?

Open the console and press one time on each button and send us the button codes.

If it's that. It will be easy to fix. Just a new model config.

**Comment by menime8** [comment]
> @menime8 Maybe it's an issue of button code. I don't remember. Did you already check the button code in the console off the browser?
> 
> Open the console and press one time on each button and send us the button codes.
> 
> If it's that. It will be easy to fix. Just a new model config.
> 
> Look at #337

I'm on it. I used the master software, put in the new pinout. Currently i try the codes

**Comment by menime8** [comment]
> > @menime8 Maybe it's an issue of button code. I don't remember. Did you already check the button code in the console off the browser?
> > Open the console and press one time on each button and send us the button codes.
> > If it's that. It will be easy to fix. Just a new model config.
> > Look at #337
> 
> I'm on it. I used the master software, put in the new pinout. Currently i try the codes

I have to use the Miami Code

**Comment by SigmaPic** [comment]
> I have to use the Miami Code 


You have tested and it fully works with Miami code ?

**Comment by Bischof-Mak** [comment]
> > You professionals have made it !!!!!
> > now I managed to load the DEV_Branch.
> > Module connected, power plugged in, everything boots up, everything can be operated (display and web interface).
> > Well done.
> 
> I'm very happy that it works correctly. Same setup that the photo you sent yesterday?
> 
> I can update the schematic and the PCB.

Yes, it is.

Tomorrow i will do it all on my BWC_V2 and after all i will give you an report..

Thanks too @SigmaPic, @visualapproach and all other for this great work. 
I am not a specialist but can help myself and with such patience and knowledge from you it was a pleasure !

**Comment by MK89OH** [comment]
Work the new PIN without cap too?

**Comment by menime8** [comment]
> Work the new PIN by the Pump s200102 without cap too?

Try it, no one can be sure to say it works or not.



@SigmaPic 
I have tested it complete now, and it works now perfect. Used mosfet, no resistors and the new pinout. 

So now we can close this issue.

Model 24949, square, works with mosfet and Miami2021 Code.

**Comment by Bischof-Mak** [comment]
> > Work the new PIN by the Pump s200102 without cap too?
> 
> Try it, no one can be sure to say it works or not.
> 
> @SigmaPic I have tested it complete now, and it works now perfect. Used mosfet, no resistors and the new pinout.
> 
> So now we can close this issue.
> 
> Model 24949, square, works with mosfet and Miami2021 Code.

Yes, it works for me with and without CAP between 5V - GND
Hawaii Hydrojet Pro - Square Pump S200102 - MOSFET (red one) - without any resistors (long cable :-))

**Comment by visualapproach** [comment]
Yay!! Everybody happy. Big thanks to all involved, especially to @SigmaPic for sharing the knowledge about electronics and figuring out the best pin out!

**Comment by menime8** [comment]
> Yay!! Everybody happy. Big thanks to all involved, especially to @SigmaPic for sharing the knowledge about electronics and figuring out the best pin out!

So now, If you're interested, my model and which code use, you see 2 posts before. Can put it to the compatible list. The 100k coffee is mine 👋

**Comment by visualapproach** [comment]
Awesome @menime8 that is really appreciated! Wish you the best and thanks again

**Comment by SigmaPic** [comment]
Nice ! I'm happy that we finally get it work.

I think that the new pinout can be used on all 6 wires models. I'm pretty sure it will work while the DSP chip is the same.

@visualapproach Do you think it's better to keep a double footprint on the PCB_V2B ?

I wonder if it's really necessary to update the PCB. It works well like that. You just need to solder the wires differently.

Maybe it's enough to explain that in the PDF.
What do you think?

**Comment by visualapproach** [comment]
@SigmaPic We can keep the PCB as is as far as I'm concerned. I like the solid ground plane that will be broken by twisted traces on a 2 layer board. A bit disturbing to twist the cables from a OCD point of view but it's probably the best way to go. I'll edit the docs and maybe remove PCB_V2 from the code? Rename B-version to just 2.

**Comment by SigmaPic** [comment]
@visualapproach 

Not sure that some traces on the bottom will break the benefits of the ground plane compared to the twisted cables. I'm not an routing expert.
But a new PCB will disturb people I think.

I think it's better to keep the PCB_V2 in the code for all the people that want to update the software without changing the hardware.

So maybe the best is just to document it in the PDF. Up to you.

**Comment by visualapproach** [comment]
@SigmaPic yeah I'm with you. Better PDF will do fine. Have a good evening now. Just enjoyed a few cold ones this tropical night at 58N

**Comment by SigmaPic** [comment]
@visualapproach 
Thanks. Have a good evening too.
Tonight I will be lazy in the spa 😉

@menime8 @Bischof-Mak 
How is the sound ? Same as original? Better ?

Maybe I can share my PCB fork. I have switched to SMD components and I have added a jumper allowing to route the original sound signal to the DSP.

**Comment by menime8** [comment]
> @visualapproach Thanks. Have a good evening too. Tonight I will be lazy in the spa 😉
> 
> @menime8 @Bischof-Mak How is the sound ? Same as original? Better ?
> 
> Maybe I can share my PCB fork. I have switched to SMD components and I have added a jumper allowing to route the original sound signal to the DSP.

Sound is ok.

**Comment by SigmaPic** [comment]
@menime8 @Bischof-Mak  another question, the display is not flashing at startup ? At startup, it behave as original ?

**Comment by SigmaPic** [comment]
Yes if it doesn't behave as original

**Comment by Bischof-Mak** [comment]
> @visualapproach Thanks. Have a good evening too. Tonight I will be lazy in the spa 😉
> 
> @menime8 @Bischof-Mak How is the sound ? Same as original? Better ?
> 
> Maybe I can share my PCB fork. I have switched to SMD components and I have added a jumper allowing to route the original sound signal to the DSP.

Hey @SigmaPic, 
By sound, do you mean the piping on the display when you press a button?

I think that's ok, you can turn it off if not desired, but do not see any difference from before. With before I mean with module. In the original is not.


> @menime8 @Bischof-Mak another question, the display is not flashing at startup ? At startup, it behave as original ?

I have no flashing at startup. Difference at startup to the original is that the date etc. scrolls...

**Comment by Bischof-Mak** [comment]
This is my finally work
![IMG_6374](https://user-images.githubusercontent.com/38432207/180376141-6d29fd6f-8302-4223-bf7e-085d3c7ddd06.jpg)
and i have translated the hmtl files to german
![Lay-Z-Spa - Krotter`s Edition](https://user-images.githubusercontent.com/38432207/180376238-5d3428b3-c4bb-458f-a123-5755fb348e15.png)
.

**Comment by menime8** [comment]
> Yes if it doesn't behave as original
Its original

**Comment by SigmaPic** [comment]
> Hey @SigmaPic, By sound, do you mean the piping on the display when you press a button?
> 
> I think that's ok, you can turn it off if not desired, but do not see any difference from before. With before I mean with module. In the original is not.
> 

Yes, I mean the bip when you press a button.
I would like to know if there a differnce between original sound (no esp => pump out of the box) and now with the PCB_V2B.

I didn't understand what you want to say with that sentence : "In the original is not."

I ask the question because the sound signal is managed a bit differently on the ESP board and I wonder if the LLC has an impact on the quality of the sound.

Maybe the difference is because the ESP doesn't produce the same melody.


> > Yes if it doesn't behave as original
> > Its original

You want to say that you have no sound difference ?

> I have no flashing at startup. Difference at startup to the original is that the date etc. scrolls...

That 's nice.

**Comment by menime8** [comment]
I noticed sometimes current action stops (pump/air) for a few seconds. It looks like the esp restarts, also the totals skip back to older values. I registered it now 2 times. 
But the timers for filter/chlor have the right values after that.

**Comment by SigmaPic** [comment]
uptime counter has been reset ?

On the bottom of SPA Config page you have boot info.
Please check and report that as well next time it happens.

**Comment by menime8** [comment]
> uptime counter has been reset ?
> 
> On the bottom of SPA Config page you have boot info. Please check and report that as well next time it happens.

Hardware Watchdog

![IMG_20220724_181412](https://user-images.githubusercontent.com/109034011/180656284-7aa0b60c-cf50-416f-be1f-55f9e26f7792.jpg)

**Comment by SigmaPic** [comment]
Perfect !

This seems to be a software bug.

@visualapproach Maybe you will be faster than me to analyse that.

I really not sure but isn't link to the wifi power ?
@visualapproach know more about that

**Comment by menime8** [comment]
I know my WiFi isn't good. Can this be the problem?

**Comment by SigmaPic** [comment]
Yes I think so, I have read it in the Issue or Discussion section. Idon't remember.
I think software timeout can be adjusted to solve the issue.

Let's see what @visualapproach say

**Comment by visualapproach** [comment]
There are two dogs. A hardware and a software one. The software barks first: after a few seconds of not getting fed. But we can put him to sleep in code so we don't have to worry about him. The other beast though will bark after not getting fed for about 7 seconds (give or take a few)
So if some process is taking more than that without passing control to the operating system the ESP will reboot.
There was a problem with websockets that could stall for 10 secs if bad Wi-Fi. But that's been taken care of by using an asynchronous server. Or so I thought. I am convinced that it is caused by bad Wi-Fi but probably something else but WS. You could try de-activating MQTT and see if it makes any difference. Or websockets.

**Comment by SigmaPic** [comment]
I just looked at mine and I also get watchdog reset. MQTT is disabled.

One is on desk, not in the pump.

![Screenshot_20220724-190429.png](https://user-images.githubusercontent.com/2431226/180658390-468550c6-b6ce-4dc1-81ec-efa7b40c8576.png)

**Comment by menime8** [comment]
I don't use MQTT

**Comment by visualapproach** [comment]
One less thing to worry about then which is good. @SigmaPic is your Wi-Fi good on the bench?

**Comment by SigmaPic** [comment]
> One less thing to worry about then which is good. @SigmaPic is your Wi-Fi good on the bench?

I think son 1 meter from the router. It cannot be better.

**Comment by visualapproach** [comment]
Could be anything I guess. If it is websockets it should be testable. Let's say you see HW WD trigger every 6 hours, or when you  shut down computer from an open esp web page or similar. Close all connections to ESP (pages in mobile phone and computers)
Wait say 12 hours or so and check last reboot.
Or do it the other way: provoke it by opening several connections and close/open and see if that triggers it.
There is also a way to use the exception decoder filter to see what it was doing at reboot, but I'm no expert on that.

**Comment by SigmaPic** [comment]
Reproduced. I opened two web page on my phone. Then give some water to plants. Ten minutes later I refresh one of the web page and I got a new watchdog right reset

**Comment by visualapproach** [comment]
Did the same. Cannot reproduce. I didn't water the plants though. You think it matters?

**Comment by SigmaPic** [comment]
> Did the same. Cannot reproduce. I didn't water the plants though. You think it matters?

Who know 🤣

I was in the garden, far from the wifi router. I the phone keep the connection but take time to answer.... 

I didn't reproduced it as well.

**Comment by SigmaPic** [comment]
I got a new wdg but I don't know how.

**Comment by SigmaPic** [comment]
@visualapproach where can we wait too long in the code ?

Is it in your code or inside a library?

Where is managed the watchdog?

**Comment by SigmaPic** [comment]
To reproduce it, I do the following:
1. Open 3 tab on the main page and 1 tab on config page on my phone with chrome 
2. Lock the phone
3. Wait 10 minutes
4. Refresh the config page => watchdog

I let it connected to the computer with a terminal opened. So I have seen the reboot. But nothing else that the boot message. No trace stack or so on. If you want to instrument the code with some print for debug, I can test it.

**Comment by visualapproach** [comment]
Testet booth on iPhone and chrome on a tablet. It works like a charm. Cannot reproduce the watchdog reset. So what's the difference ? Router is probably different between us, and the pin out. I use PCB V2 as is. I doubt that the pins should be the reason for a wd timeout.
I have deco routers. What's yours @SigmaPic @menime8 ?

**Comment by visualapproach** [comment]
My bootlog.txt, (can be found in file system on the esp)

{"BOOTINFO":"Power On 2022-06-02 19:19:35"}
{"BOOTINFO":"Power On 2022-06-02 19:20:31"}
{"BOOTINFO":"Software/System restart 2022-06-04 16:00:09"}
{"BOOTINFO":"External System 2022-06-26 22:08:19"}
{"BOOTINFO":"Software/System restart 2022-06-26 22:20:41"}
{"BOOTINFO":"Software/System restart 2022-06-26 22:34:09"}
{"BOOTINFO":"Software/System restart 2022-06-28 09:19:43"}
{"BOOTINFO":"External System 2022-06-28 10:16:50"}
{"BOOTINFO":"Power On 2022-06-28 10:17:49"}
{"BOOTINFO":"External System 2022-06-28 15:14:12"}
{"BOOTINFO":"Exception 2022-06-28 15:15:04"}
{"BOOTINFO":"External System 2022-06-28 15:16:30"}
{"BOOTINFO":"Exception 2022-06-28 15:19:15"}
{"BOOTINFO":"External System 2022-06-28 15:19:47"}
{"BOOTINFO":"External System 2022-06-28 15:20:06"}
{"BOOTINFO":"Software/System restart 2022-06-28 15:23:35"}
{"BOOTINFO":"Power On 2022-06-28 15:24:19"}
{"BOOTINFO":"Software/System restart 2022-06-28 15:26:39"}
{"BOOTINFO":"Software/System restart 2022-06-29 16:24:25"}
{"BOOTINFO":"External System 2022-06-29 17:38:58"}
{"BOOTINFO":"External System 2022-06-29 17:39:17"}
{"BOOTINFO":"Power On 2022-06-29 17:39:54"}
{"BOOTINFO":"Power On 2022-06-29 20:15:12"}
{"BOOTINFO":"Software/System restart 2022-06-29 20:17:30"}
{"BOOTINFO":"Software/System restart 2022-06-29 20:48:08"}
{"BOOTINFO":"Software/System restart 2022-07-03 09:38:30"}
{"BOOTINFO":"Software/System restart 2022-07-03 09:59:01"}
{"BOOTINFO":"Software/System restart 2022-07-03 10:06:37"}
{"BOOTINFO":"Software/System restart 2022-07-03 10:28:54"}
{"BOOTINFO":"Software/System restart 2022-07-03 10:32:31"}
{"BOOTINFO":"Software/System restart 2022-07-03 10:41:56"}
{"BOOTINFO":"Software/System restart 2022-07-03 11:12:17"}
{"BOOTINFO":"External System 2022-07-03 11:12:53"}
{"BOOTINFO":"External System 2022-07-03 11:15:39"}
{"BOOTINFO":"Software/System restart 2022-07-03 11:17:43"}
{"BOOTINFO":"Software/System restart 2022-07-03 11:21:21"}
{"BOOTINFO":"Power On 2022-07-03 11:25:52"}
{"BOOTINFO":"Exception 2022-07-03 11:26:10"}
{"BOOTINFO":"External System 2022-07-03 11:26:34"}
{"BOOTINFO":"External System 2022-07-03 11:27:25"}
{"BOOTINFO":"External System 2022-07-03 11:27:35"}
{"BOOTINFO":"Software/System restart 2022-07-03 11:29:46"}
{"BOOTINFO":"Power On 2022-07-03 11:30:33"}
{"BOOTINFO":"Software/System restart 2022-07-03 12:12:26"}
{"BOOTINFO":"Software/System restart 2022-07-03 12:26:09"}
{"BOOTINFO":"Software/System restart 2022-07-03 12:30:44"}
{"BOOTINFO":"Software/System restart 2022-07-08 12:02:07"}
{"BOOTINFO":"Software/System restart 2022-07-08 12:15:33"}
{"BOOTINFO":"Software/System restart 2022-07-08 13:27:05"}
{"BOOTINFO":"External System 2022-07-08 13:27:28"}
{"BOOTINFO":"External System 2022-07-08 13:28:09"}
{"BOOTINFO":"External System 2022-07-08 13:28:18"}
{"BOOTINFO":"Software/System restart 2022-07-08 13:31:17"}
{"BOOTINFO":"Software/System restart 2022-07-08 13:39:20"}
{"BOOTINFO":"Power On 2022-07-08 13:41:38"}
{"BOOTINFO":"Software/System restart 2022-07-08 13:48:21"}
{"BOOTINFO":"Software/System restart 2022-07-08 14:07:37"}
{"BOOTINFO":"Software/System restart 2022-07-09 08:19:12"}
{"BOOTINFO":"Software/System restart 2022-07-09 09:13:19"}
{"BOOTINFO":"Software/System restart 2022-07-09 09:35:13"}
{"BOOTINFO":"Software/System restart 2022-07-09 09:46:49"}
{"BOOTINFO":"Software/System restart 2022-07-09 09:52:45"}
{"BOOTINFO":"Software/System restart 2022-07-09 10:01:21"}
{"BOOTINFO":"Software/System restart 2022-07-09 10:05:17"}
{"BOOTINFO":"Software/System restart 2022-07-09 10:47:38"}
{"BOOTINFO":"External System 2022-07-09 10:50:31"}
{"BOOTINFO":"Software/System restart 2022-07-09 11:16:10"}
{"BOOTINFO":"External System 2022-07-09 11:21:14"}
{"BOOTINFO":"External System 2022-07-09 13:03:44"}
{"BOOTINFO":"Software/System restart 2022-07-09 13:06:21"}
{"BOOTINFO":"Software/System restart 2022-07-09 13:20:25"}
{"BOOTINFO":"Software/System restart 2022-07-09 13:28:45"}
{"BOOTINFO":"Software/System restart 2022-07-09 13:33:47"}
{"BOOTINFO":"Software/System restart 2022-07-09 14:10:06"}
{"BOOTINFO":"Software/System restart 2022-07-09 15:52:43"}
{"BOOTINFO":"Software/System restart 2022-07-11 16:49:50"}
{"BOOTINFO":"Software/System restart 2022-07-11 16:58:52"}
{"BOOTINFO":"Software/System restart 2022-07-13 09:44:32"}
{"BOOTINFO":"Software/System restart 2022-07-13 09:47:52"}
{"BOOTINFO":"Software/System restart 2022-07-13 09:56:43"}
{"BOOTINFO":"Software/System restart 2022-07-15 22:43:22"}
{"BOOTINFO":"Software/System restart 2022-07-16 11:38:45"}
{"BOOTINFO":"Power On 2022-07-16 11:52:26"}
{"BOOTINFO":"Power On 2022-07-16 11:53:06"}
{"BOOTINFO":"Power On 2022-07-16 11:56:24"}
{"BOOTINFO":"Software/System restart 2022-07-16 12:40:51"}
{"BOOTINFO":"Power On 2022-07-16 12:41:28"}
{"BOOTINFO":"Software/System restart 2022-07-16 12:46:06"}
{"BOOTINFO":"Software/System restart 2022-07-16 12:51:12"}
{"BOOTINFO":"Software/System restart 2022-07-16 13:04:01"}
{"BOOTINFO":"Software/System restart 2022-07-16 13:12:16"}
{"BOOTINFO":"Software/System restart 2022-07-16 13:22:43"}
{"BOOTINFO":"Software/System restart 2022-07-18 15:24:44"}
{"BOOTINFO":"Software/System restart 2022-07-18 15:34:11"}
{"BOOTINFO":"Software/System restart 2022-07-18 16:52:10"}
{"BOOTINFO":"Software/System restart 2022-07-18 17:29:08"}
{"BOOTINFO":"Software/System restart 2022-07-18 17:35:01"}
{"BOOTINFO":"Software/System restart 2022-07-19 06:11:12"}
{"BOOTINFO":"Software/System restart 2022-07-19 09:17:38"}
{"BOOTINFO":"Software/System restart 2022-07-19 09:18:47"}
{"BOOTINFO":"Software/System restart 2022-07-21 14:23:04"}
{"BOOTINFO":"Software/System restart 2022-07-21 19:08:45"}
{"BOOTINFO":"Software/System restart 2022-07-23 09:03:57"}
{"BOOTINFO":"Software/System restart 2022-07-23 13:53:22"}
{"BOOTINFO":"Software/System restart 2022-07-23 14:47:21"}
{"BOOTINFO":"Software/System restart 2022-07-23 16:04:43"}
{"BOOTINFO":"Software/System restart 2022-07-23 16:08:04"}
{"BOOTINFO":"Software/System restart 2022-07-23 16:18:10"}
{"BOOTINFO":"Software/System restart 2022-07-23 21:48:33"}

**Comment by SigmaPic** [comment]
> Testet booth on iPhone and chrome on a tablet. It works like a charm. Cannot reproduce the watchdog reset. So what's the difference ? Router is probably different between us, and the pin out. I use PCB V2 as is. I doubt that the pins should be the reason for a wd timeout.
> I have deco routers. What's yours @SigmaPic @menime8 ?

I don't have any PCB. Just the esp alone.

I have a TP Link router. I doubt that the router is the cause.

I try again this evening to reproduce.

Maybe such issue : [https://solace.community/discussion/694/client-websocket-disconnects-when-browser-is-minimized-or-not-in-focus-chrome](https://solace.community/discussion/694/client-websocket-disconnects-when-browser-is-minimized-or-not-in-focus-chrome)

**Comment by visualapproach** [comment]
Reading up a little on the AsyncWebSockets (which I should use instead. Thought I was but someone else added it and I didn't pay attention) and it says this. May be the same on the usual server...
`Limiting the number of web socket clients
Browsers sometimes do not correctly close the websocket connection, even when the close() function is called in javascript. This will eventually exhaust the web server's resources and will cause the server to crash. Periodically calling the cleanClients() function from the main loop() function limits the number of clients by closing the oldest client when the maximum number of clients has been exceeded. This can called be every cycle, however, if you wish to use less power, then calling as infrequently as once per second is sufficient.`

**Comment by visualapproach** [comment]
AsyncWS was too different to switch to and takes too much effort.
But..
What happens if you edit this function in `main.cpp` 
```
/**
 * start a web socket server
 */
void startWebSocket()
{
  // In case we are already running
  webSocket.close();
  webSocket.begin();
  webSocket.enableHeartbeat(3000, 3000, 1);  //*******insert this line**********
  webSocket.onEvent(webSocketEvent);
  Serial.println(F("WebSocket > server started"));
}
```

**Comment by SigmaPic** [comment]
> AsyncWS was too different to switch to and takes too much effort. But.. What happens if you edit this function in `main.cpp`
> 
> ```
> /**
>  * start a web socket server
>  */
> void startWebSocket()
> {
>   // In case we are already running
>   webSocket.close();
>   webSocket.begin();
>   webSocket.enableHeartbeat(3000, 3000, 1);  //*******insert this line**********
>   webSocket.onEvent(webSocketEvent);
>   Serial.println(F("WebSocket > server started"));
> }
> ```

I tried it. I will tell you if it solve the issue.

**Comment by SigmaPic** [comment]
@visualapproach Still hardware reset.

If you want to instrument the code to output on the serial line, I can test it.

**Comment by SigmaPic** [comment]
@visualapproach @menime8 what version of Android do you have ?

I have Android 12.

It may be linked to the OS.

**Comment by menime8** [comment]
> @visualapproach @menime8 what version of Android do you have ?
> 
> I have Android 12.
> 
> It may be linked to the OS.

12, too. I was the whole day at work, No reboots since yesterday evening.

**Comment by SigmaPic** [comment]
Maybe Android 12 is the reason....

For power consumption reduction, they may implement specific behavior with websockets.

@menime8 No reset without any page open ?

**Comment by visualapproach** [comment]
10

**Comment by menime8** [comment]
Yes. No page, No reset.

**Comment by visualapproach** [comment]
@SigmaPic if you want to try something you can comment out all ESP.wdtDisable() and ESP.wdtFeed() in BWC_common.cpp. If you still get HARDWARE watchdog reboots we ruled out my part of the code and its very likely to be websockets or http related.

**Comment by SigmaPic** [comment]
> @SigmaPic if you want to try something you can comment out all ESP.wdtDisable() and ESP.wdtFeed() in BWC_common.cpp. If you still get HARDWARE watchdog reboots we ruled out my part of the code and its very likely to be websockets or http related.

wdtEnable as well ?

**Comment by visualapproach** [comment]
May as well

**Comment by visualapproach** [comment]
I'm away now but there is a debug mode in websockets. See if you can figure out how to turn it on. Maybe define DEBUG or similar in main.cpp

**Comment by SigmaPic** [comment]
> I'm away now but there is a debug mode in websockets. See if you can figure out how to turn it on. Maybe define DEBUG or similar in main.cpp 

You're right, I saw the debug mode.

I have compiled without the wdt. Let's see that first. Then I will try the debug mode.

**Comment by SigmaPic** [comment]
@visualapproach cannot reproduced for the moment by removing all ESP.wdtXXX

**Comment by visualapproach** [comment]
That's good in a weird way. Do you get sw reboot if you save stuff like price/kwh?

**Comment by SigmaPic** [comment]
Still no reboot during the night. I removed the heartbeat. So the wdtXXX seems to help.

I tried to save several time the config with that SW, no reboot.

**Comment by SigmaPic** [comment]
The wdt is feed only on get/setJSONxxx.
In some other function, I'm the wdt is disabled.

It means that these functions would not called during a few time ?
I think they are used by the webserver, no ?

**Comment by SigmaPic** [comment]
@visualapproach why do you use the software watchdog ?

I have read that its timeout is not well defined, around 1.5 and 3 seconds.

Why the hardware watchdog is not enough ?

**Comment by visualapproach** [comment]
Technically I try to Not use the watchdog but I see what you mean. Espressif made it this way and the idea is good; if your program hangs while the device is doing something important and is hard to get to, it resets and can hopefully recover. That would be the HWWDT I guess. Then I think they wanted a second WDT to ensure Wi-Fi and the OS gets enough CPU time. Anyway, during development I had problems with the SWWDT triggering so I fed it or killed it and it worked. It's possible that something changed with the new OS and arduino core updates so it's not necessary to disable the sw wdt. Or that there was something else triggering.

**Comment by visualapproach** [comment]
I don't get why enabling the sw wdt would keep the hw wdt from triggering. Or to be more correct, the sw wdt is enabled all the time, we just disabled it at time consuming functions. But if these functions would stall and trigger the hw wdt we would get the sw wdt instead now. Unless some other code relies on the sw wdt. Well if it continue to work I'll update the code.

**Comment by RatAttack83** [comment]
Mine is working now, too! ☺️ I will post it here, if there is a wd problem... But I think its everything ok! ☺️ So tomorrow i will print a case and than project finish! 
Thank you everybody for the troubleshooting! ☺️

**Comment by SigmaPic** [comment]
> Technically I try to Not use the watchdog but I see what you mean. Espressif made it this way and the idea is good; if your program hangs while the device is doing something important and is hard to get to, it resets and can hopefully recover. That would be the HWWDT I guess. Then I think they wanted a second WDT to ensure Wi-Fi and the OS gets enough CPU time. Anyway, during development I had problems with the SWWDT triggering so I fed it or killed it and it worked. It's possible that something changed with the new OS and arduino core updates so it's not necessary to disable the sw wdt. Or that there was something else triggering.

I'm not sure to understand what you mean and what has been imagined by expressif.

If I'm right, hwdt timeout cannot be changed but swdt can (for the moment it seems not possible for the swdt too).

When you feed the swdt, it feed the hwdt as well. 

My understanding is that swdt is just the way to adjust the timeout of the watchdog but currently it's not possible.

What is strange is that they are linked. When you feed the swdt, it feed the hwdt. This is unlogical. 

It could be interesting if you could setup several swdt and in that case hwdt would only be fed if all swdt are fed.

For this project, if hwdt timeout is ok (6s if I'm right) then we can only work with hwdt. If swdt timeout is needed (less than 6s), you could eventually enable it for some parts of the code but I don't see the reason why a reduced timeout would be needed for us.

So, I think we can disable the swdt for all the project but we can still feed the hwdt in BWC_common.cpp if needed.

The best would be to implement two swdt so that you can manage two timeouts, one for the webserver part for example, one for the MQTT... But one hwdt is enough I think. That's already a good practice to have one hwdt.

**Comment by SigmaPic** [comment]
> I don't get why enabling the sw wdt would keep the hw wdt from triggering. Or to be more correct, the sw wdt is enabled all the time, we just disabled it at time consuming functions. But if these functions would stall and trigger the hw wdt we would get the sw wdt instead now. Unless some other code relies on the sw wdt. Well if it continue to work I'll update the code.

I explain before, enabling the sw wdt don't keep the hw wdt from triggering. It's feeding the swdt that feed the hw wdt as well. So hw wdt won't be fired it swdt is feed correctly because swdt timeout is shorter.


I still don't have any watchdog reboot since I removed all the wdt calls in BWC_common.cpp.

If I'm right, in some functions m, you disable the swdt when you enter the function and you enable it when you exit. This means that swdt is enabled for the rest of the execution of the software, thus for the websockets for example. 

Why I trying to say is that you have the swdt timeout for the websockets and maybe it's not enough, especially when you have several clients.

Maybe I'm speculating 😂

**Comment by visualapproach** [comment]
@SigmaPic 
> I'm not sure to understand what you mean and what has been imagined by expressif.

Something got lost in translation I think. Don't know how to answer that. Moving on.

But it's clear that the documentation on the K9's are unclear...

> So, I think we can disable the swdt for all the project but we can still feed the hwdt in BWC_common.cpp if needed.

Possibly, but it's there to guard against full receive buffers and stuff, so then we would get crazy errors instead of a sw reset. I'm also speculating here. But If it works for you guys I think I'll remove the wdt-stuff.

**Comment by SigmaPic** [comment]
@menime8 can you confirm it works for you as well. 

To check, just remove all lines that start with "ESP.wdt" in BWC_common.cpp, rebuild and check.

**Comment by menime8** [comment]
> @menime8 can you confirm it works for you as well.
> 
> To check, just remove all lines that start with "ESP.wdt" in BWC_common.cpp, rebuild and check.

Never Touch a running system 😁

Last reboot was at July, 24th. The problem is too much tabs in my explorer. So i can handle it now. If i have time next week, i try it.

**Comment by visualapproach** [comment]
@SigmaPic I experimented a bit with the dogs. When I removed the statements that disabled and fed the dog I too had hardware watchdog resets when I had one page open on computer and one on iPhone. When I locked the screen on iPhone I got hw wdt every time. Now I have ESP.wdtFeed() in all those functions, and it seems okay again. Like you said :-)

**Comment by SigmaPic** [comment]
If all the functions are called within the main main loop, you can feed the dog within them.

Don't feed it in interrupts because you will loose all the benefits of the watchdog.

**Comment by SigmaPic** [comment]
On my side I got all the components. I solder everything and it was plug and play. It works great.

What is the best to perform an OTA ?
Power off the pump ?

**Comment by visualapproach** [comment]
Yes I fed the dog in the functions called by main loop indirectly. Interrupts should be as short as possible and everything must be in iram so I didn't plan to do that 😃 

I don't fully understand your question about OTA. You can do it anytime the module is online. I do it when it's hidden in the pump as well as on the bench. Maybe avoid rainy days in case it doesn't work so you don't have to go outside.

**Comment by SigmaPic** [comment]
Ok for the 🐶 😀

I wonder what happens if the ESP stops working during the OTA whereas the pump continue to send SPI frames.

All the pins are in input during OTA ?

**Comment by SigmaPic** [comment]
And also, is there a power button on the web page ? The same button you have on the pump ?

**Comment by visualapproach** [comment]
> I wonder what happens if the ESP stops working during the OTA whereas the pump continue to send SPI frames.
> 
> All the pins are in input during OTA ?

Not sure about the pin mode during the OTA, but I have done it many times on PCB V1 and 2 without any problems. The 4 wire systems are less tolerant than 6 wire. They get E13 com failure after a few seconds, but the 6 wire has no problem with that.

**Comment by visualapproach** [comment]
> And also, is there a power button on the web page ? The same button you have on the pump ?
> 

Nope, I deliberately removed that functionality. As far as I know the only thing it did was dimming the display. Which you can do in the web ui anyway. Perhaps it put the cpu in sleep mode to save half a watt or so, but I doubt it.

Technically it's not difficult to expose it but why add an extra useless button press?

**Comment by SigmaPic** [comment]
I see two reasons:
- when you want to power off the pump, you can do it through one command
- when you look at the pump you don't have the green led ON.

**Comment by RobReus** [comment]
Commenting here since I did not see it being mentioned anywhere else.

I built the controller for my 2021 Maldives (6-wire S100104 square pump) using the PCB_V2B wiring. Instead of using the red LLC, I used the blue LLC (TXS0108E) and it worked like a charm. I did have to add resistors to the display output (D3, D4 and D6 iirc) to stop the flashing, but other than that, it worked flawlessly.

I forgot to take pictures, but the wiring looked identical to what has been posted in this issue already, except for the fact that those 3 wires for the display are not connected to the A pins but to the B pins with a 560Ohm resistor.

Great job to everyone involved on troubleshooting, debugging, and finding a working solution for these new models!
